tree/readspeed/src/ReadSpeed.cxx

```diff
@@ -0,0 +1,314 @@
+#include <TTree.h>
+
+#include <algorithm>
+#include <cassert>
+#include <cmath> // std::ceil
+#include <memory>
+#include <stdexcept>
+#include <set>
+#include <regex>
+
+using namespace ReadSpeed;
+
+std::vector<std::string> ReadSpeed::GetMatchingBranchNames(const std::string &fileName, const std::string &treeName,
+                                                           const std::vector<std::string> &regexes)
+{
+   TFile *f = TFile::Open(fileName.c_str(), "READ_WITHOUT_GLOBALREGISTRATION");
+   if (f == nullptr || f->IsZombie())
+      throw std::runtime_error("Could not open file '" + fileName + '\'');
+   std::unique_ptr<TTree> t(f->Get<TTree>(treeName.c_str()));
+   if (t == nullptr)
+      throw std::runtime_error("Could not retrieve tree '" + treeName + "' from file '" + fileName + '\'');
+
+   const auto unfilteredBranchNames = ROOT::Internal::TreeUtils::GetTopLevelBranchNames(*t);
+   std::set<std::string> usedRegexes;
+   std::vector<std::string> branchNames;
+
+   auto filterBranchName = [regexes, &usedRegexes](std::string bName) {
+      const auto matchBranch = [&usedRegexes, bName](std::string regex) {
+         std::regex branchRegex(regex);
+         bool match = std::regex_match(bName, branchRegex);
+
+         if (match)
+            usedRegexes.insert(regex);
+
+         return match;
+      };
+
+      const auto iterator = std::find_if(regexes.begin(), regexes.end(), matchBranch);
+      return iterator != regexes.end();
+   };
+   std::copy_if(unfilteredBranchNames.begin(), unfilteredBranchNames.end(), std::back_inserter(branchNames),
+                filterBranchName);
+
+   if (branchNames.empty())
+      throw std::runtime_error("Provided branch regexes didn't match any branches in the tree.");
+   if (usedRegexes.size() != regexes.size()) {
+      std::string errString =
+         "The following regexes didn't match any branches in the tree, this is probably unintended:\n";
+      for (const auto &regex : regexes) {
+         if (usedRegexes.find(regex) == usedRegexes.end())
+            errString += '\t' + regex + '\n';
+      }
+      throw std::runtime_error(errString);
+   }
+
+   return branchNames;
+}
+
+// Read branches listed in branchNames in tree treeName in file fileName, return number of uncompressed bytes read.
+ByteData ReadSpeed::ReadTree(const std::string &treeName, const std::string &fileName,
+                             const std::vector<std::string> &branchNames, EntryRange range)
+{
+   // This logic avoids re-opening the same file many times if not needed
+   // Given the static lifetime of `f`, we cannot use a `unique_ptr<TFile>` lest we have issues at teardown
+   // (e.g. because this file outlives ROOT global lists). Instead we rely on ROOT's memory management.
+   thread_local TFile *f;
+   if (f == nullptr || f->GetName() != fileName) {
+      delete f;
+      f = TFile::Open(fileName.c_str(), "READ_WITHOUT_GLOBALREGISTRATION"); // TFile::Open uses plug-ins if needed
+   }
```

Comment:
I just realized that "we rely on ROOT's memory management" does not really work with `"READ_WITHOUT_GLOBALREGISTRATION"`. But we really need `READ_WITHOUT_GLOBALREGISTRATION` to avoid contention between threads. Calling `delete f` at the end of this function is a possible performance pessimization because it makes it impossible for subsequent calls to `ReadTree` to re-use a previously-opened `TFile`.

Any ideas how to `READ_WITHOUT_GLOBALREGISTRATION` _and_ reuse `TFile`s whenever possible _and_ not leave the last files processed open? :smile:

Reply:
We could add the file pointer as an argument to read tree and open it before we call the function? Still avoids registration and re-opening and lets us just close it after the function call

Reply:
correct, and we need one TFile pointer (better a `unique_ptr<TFile>` _per thread_). Then threads can decide who gets which file pointer based, for example, on an index pulled from a [thread-safe stack of indexes](https://github.com/root-project/root/blob/master/tree/dataframe/inc/ROOT/RDF/RSlotStack.hxx) (or if you can think of a way to do it without synchronization between threads required, even better).

Reply:
Okay file access should now only involve opening and closing a file a single time per thread now. Let me know if it's too convoluted or something!